mkosi/__init__.py

```diff
@@ -4498,6 +4507,15 @@ def parse_base_packages(value: str) -> Union[str, bool]:
     return parse_boolean(value)
 
 
+def parse_remove_files(value: str) -> List[str]:
+    """Normalize paths as relative to / to ensure we don't go outside of our root."""
+
+    # os.path.normpath() doesn't normalize exactly '//' (prolly
+    # inspired by Windows paths), even though it normalizes '///',
+    # so we can't just prepend '/' unconditionally.
```

Comment:
This sent me down a little rabbit hole (after looking at the implementation of `os.path.normpath`) and TIL. :) This is actually [POSIX fault](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13)
> If a pathname begins with two successive <slash> characters, the first component following the leading <slash> characters may be interpreted in an implementation-defined manner, although more than two leading <slash> characters shall be treated as a single <slash> character.

I haven't grepped through the kernel code, but quickly browsing the Linux Programming Interface I only find the known rule that multiple slashes are collapsed into one, implying that initial slashes are not different from later ones and this behaviour `normpath` being for compatibility with other weird unices (apparently cygwin treats them differently). 

Another quick check shows me that at least `os.unlink` doesn't care about `//foo` vs `/foo` (as it should, since the underlying libc and syscall probably won't), so I guess you *can* just unconditionally prepend a slash.

Reply:
Thanks for the correction. I'll adjust the comment to blame POSIX instead of MS.



> Another quick check shows me that at least `os.unlink` doesn't care about `//foo` vs `/foo` (as it should, since the underlying libc and syscall probably won't), so I guess you _can_ just unconditionally prepend a slash.

The syscalls don't care, but `os.path.join()` does. And also `pathlib.Path.__div__()` does, which I was very disappointed to see. It makes it very annoying to use either of the two APIs, because you have to constantly check if the path does not start with a slash or do other workarounds.

Duh, but now I see that my code is borked, because it'll mishandle a user-specified path starting with '//'. I'll rework this.

Reply:
As a side note for the problem of reparenting a path with pathlib `Path`s, I use this as an idiom `newroot / oldroot.relative_to(path)`.

Reply:
That looked promising… but both paths must be absolute, or both must be relative. So it's not useful for the case here. I think the approach with `.lstrip()` is OK. Maybe not very elegant, but it's simple and should be very robust.